File: 1971_Find if Path Exists in Graph/FindifPathExistsinGraph_DFS.cs
public class Solution {
    private bool seen = false;
    public bool ValidPath(int n, int[][] edges, int start, int end) {
        if(start == end) return true;
        
        var graph = new Dictionary<int, List<int>>();
        
        for (int i = 0; i < edges.Length; i++)
        {
            var pair = edges[i];

            if (!graph.ContainsKey(pair[0])) graph.Add(pair[0], new List<int>());
            if (!graph.ContainsKey(pair[1])) graph.Add(pair[1], new List<int>());

            graph[pair[0]].Add(pair[1]);
            graph[pair[1]].Add(pair[0]);
        }
        
        var visited = new HashSet<int>();
        DFS(graph, start, end, visited);
        return seen;
    }
    
    private void DFS(Dictionary<int, List<int>> graph, int start, int end, HashSet<int> visited)
    {
        if(seen) return;
        
        if(start == end)
        {
            seen = true;
            return;
        }
        
        if(visited.Contains(start)) return;
        visited.Add(start);
        foreach(var next in graph[start])
        {
            DFS(graph, next, end, visited);
        }
    }
}

File: 1971_Find if Path Exists in Graph/FindifPathExistsinGraph_DFS.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        if(source == destination) return true;
        Map<Integer,Set<Integer>> graph = build(edges);
        return traverse(graph, source, destination, new HashSet<Integer>());
    }

    private boolean traverse(Map<Integer, Set<Integer>> graph, int source, int desc, Set<Integer> visited) {
        if(visited.contains(source) || !graph.containsKey(source)) return false;
        if(source == desc) return true;
        visited.add(source);
        for(int next : graph.get(source)) {
            if(traverse(graph, next, desc, visited)) return true;
        }
        return false;

    }

    private Map<Integer, Set<Integer>> build(int[][] edges) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        for(int[] edge : edges) {
            int vertexA =edge[0], vertexB = edge[1];
            graph.putIfAbsent(vertexA, new HashSet<Integer>());
            graph.putIfAbsent(vertexB, new HashSet<Integer>());
            graph.get(vertexA).add(vertexB);
            graph.get(vertexB).add(vertexA);
        }
        return graph;
    }
}